specifyweb/frontend/js_src/lib/components/Attachments/UploadStateDialog.tsx

```diff
@@ -0,0 +1,476 @@
+                  handleUploadReMap(undefined);
+                }}
+              >
+                {attachmentsText.beginUploadDescription()}
+              </Dialog>
+            ) : (
+              <Dialog
+                buttons={
+                  <Button.DialogClose>{commonText.close()}</Button.DialogClose>
+                }
+                header={attachmentsText.attachmentServerUnavailable()}
+                icon={dialogIcons.warning}
+                onClose={() => {
+                  handleSync(undefined, false);
+                  setTriedUpload('base');
+                }}
+              >
+                <p>{attachmentsText.attachmentServerUnavailable()}</p>
+              </Dialog>
+            )
+          }
+        </AttachmentsAvailable>
+      )}
+
+      {
+        /*
+         * If upload was confirmed, but dataset status hasn't been set to uploading,
+         * the uploader is validating, and generating tokens. Display loading screen
+         * in that case
+         */
+        upload === 'confirmed' && dataSet.status !== 'uploading' && (
+          <LoadingScreen />
+        )
+      }
+    </>
+  );
+}
+
+function UploadState({
+  workProgress,
+  workRef,
+  onStop: handleStop,
+  onCompletedWork: handleCompletedWork,
+  triggerNow,
+}: AttachmentWorkStateProps<'uploading'>): JSX.Element | null {
+  return workProgress.type === 'safe' ? (
+    <Dialog
+      buttons={
+        <Button.Danger onClick={handleStop}>{wbText.stop()}</Button.Danger>
+      }
+      header={wbText.uploading()}
+      onClose={() => undefined}
+    >
+      {attachmentsText.filesUploaded({
+        uploaded: workProgress.uploaded,
+        total: workProgress.total,
+      })}
+      <Progress max={workProgress.total} value={workProgress.uploaded} />
+    </Dialog>
+  ) : workProgress.type === 'stopping' ? (
+    <Dialog
+      buttons={<></>}
+      header={wbText.aborting()}
+      onClose={() => undefined}
+    >
+      {wbText.aborting()}
+    </Dialog>
+  ) : workProgress.type === 'stopped' ? (
+    <Dialog
+      buttons={<Button.DialogClose>{commonText.close()}</Button.DialogClose>}
+      header={wbText.uploadCanceled()}
+      onClose={() => handleCompletedWork(workRef.current.mappedFiles)}
+    >
+      {wbText.uploadCanceledDescription()}
+    </Dialog>
+  ) : workProgress.type === 'interrupted' ? (
+    <Dialog
+      buttons={
+        <>
+          <Button.Danger onClick={handleStop}>{wbText.stop()}</Button.Danger>
+          <Button.Fancy onClick={triggerNow}>
+            {attachmentsText.tryNow()}
+          </Button.Fancy>
+        </>
+      }
+      header={attachmentsText.interrupted()}
+      onClose={() => undefined}
+    >
+      {attachmentsText.interruptedTime({
+        remainingTime: formatTime(workProgress.retryingIn),
+      })}
+    </Dialog>
+  ) : null;
+}
+
+export async function getFileValidity(file: File): Promise<void> {
+  /*
+   * Read the first byte of the file. If this fails, it would mean
```

Comment:
I added this, in case the user deletes the file from their computer after selecting it, but before clicking on upload. I added the above check, to test for that -- but removed that (thus the early return) because I'm doing testing on files with size 0 (on which it fails)

Is this case worth handling it? I'm actually not sure how common that case will be. My next commit will remove that check entirely, so it will handle this case by cancelling the entire upload progress (worse than before, as the above will skip just that file), but I don't know if it is worth handing.